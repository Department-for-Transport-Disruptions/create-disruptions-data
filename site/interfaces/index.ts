import { NextApiRequest } from "next";
import { z } from "zod";
import { ServerResponse } from "http";
import { Service, Stop } from "../schemas/consequence.schema";
import { ConsequenceType } from "../schemas/type-of-consequence.schema";

export interface ErrorInfo {
    errorMessage: string;
    id: string;
    userInput?: string;
}

export interface ResponseWithLocals extends ServerResponse {
    locals: {
        nonce: string;
        csrfToken: string;
    };
}

export interface CookiePolicy {
    essential: boolean;
    usage: boolean;
}

export interface CookiesApiRequest extends NextApiRequest {
    body: {
        tracking: "on" | "off";
    };
}

export interface FormBase<T> {
    value?: string;
    display: string;
    displaySize?: "s" | "m" | "l" | "xl";
    inputName: Extract<keyof T, string>;
    initialErrors?: ErrorInfo[];
    stateUpdater: (change: string, field: keyof T) => void;
    schema?: z.ZodTypeAny;
}

export interface DisplayValuePair<T = string> {
    display: string;
    value: T;
    checked?: boolean;
}

export interface Feedback {
    question: string;
    answer: string;
}

export interface TestInputs {
    field1: string;
    field2: boolean;
    field3: string;
    field4: string;
}

export interface PageState<T> {
    errors: ErrorInfo[];
    inputs: T;
}

export interface SocialMediaPost {
    messageToAppear: string;
    publishDate: string;
    publishTime: string;
    accountToPublish: string;
}

export interface CreateConsequenceProps<T> {
    initialPageState: PageState<Partial<T>>;
    previousConsequenceInformation: ConsequenceType;
    initialServices?: Service[];
    initialStops?: Stop[];
    allStops?: Stop[]
}
