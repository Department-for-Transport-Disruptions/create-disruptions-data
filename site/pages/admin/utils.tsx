import resendInvite from "../api/admin/resend-invite.api";

const createUserActionsLink = (
    key: string,
    index: number,
    username: string,
    userGroup: string,
    showResendInvite?: boolean,
) => {
    return (
        <>
            {showResendInvite ? (
                <>
                    <button
                        key={`${key}${index ? `-${index}` : ""}`}
                        className="govuk-link"
                        onClick={() => resendInvite(username, userGroup)}
                    >
                        Resend invite
                    </button>
                    <br />
                    <button
                        key={`${key}${index ? `-remove-${index}` : "-remove"}`}
                        className="govuk-link"
                        onClick={() => removeUser(username)}
                    >
                        Remove
                    </button>
                </>
            ) : (
                <button
                    key={`${key}${index ? `-${index}` : ""}`}
                    className="govuk-link"
                    onClick={() => removeUser(username)}
                >
                    Remove
                </button>
            )}
        </>
    );
};
