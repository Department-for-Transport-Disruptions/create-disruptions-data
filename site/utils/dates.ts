import dayjs from "dayjs";
import customParseFormat from "dayjs/plugin/customParseFormat";
import isBetween from "dayjs/plugin/isBetween";
import isSameOrAfter from "dayjs/plugin/isSameOrAfter";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore";
import { CD_DATE_FORMAT } from "../constants";

dayjs.extend(customParseFormat);
dayjs.extend(isSameOrAfter);
dayjs.extend(isSameOrBefore);
dayjs.extend(isBetween);

export const convertDateTimeToFormat = (dateOrTime: string | Date, format: string = CD_DATE_FORMAT) =>
    dayjs(dateOrTime).format(format);

export const getDate = (input?: string) => dayjs(!!input ? input : undefined);

export const getFormattedDate = (date: string | Date) => dayjs(date, "DD/MM/YYYY");

export const formatTime = (time: string) => (time.length === 4 ? time.slice(0, -2) + ":" + time.slice(-2) : time);

export const getDatetimeFromDateAndTime = (date: string, time: string) => dayjs(`${date} ${time}`, "DD/MM/YYYY HHmm");

export const getFutureDateAsString = (addDays: number, dateFormat = CD_DATE_FORMAT) => {
    return dayjs().add(addDays, "day").format(dateFormat).toString();
};
export const checkOverlap = (firstStartDate: dayjs.Dayjs, firstEndDate: dayjs.Dayjs, secondStartDate: dayjs.Dayjs) => {
    return (
        firstStartDate.isBetween(secondStartDate, firstEndDate) ||
        secondStartDate.isBetween(firstStartDate, firstEndDate) ||
        firstStartDate.isSame(secondStartDate)
    );
};
