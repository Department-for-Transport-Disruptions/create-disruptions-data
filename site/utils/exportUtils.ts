import { Schema } from "write-excel-file";
import { ExportDisruptionData } from "../schemas/disruption.schema";

export const getExportSchema = (): Schema<ExportDisruptionData> => {
    return [
        {
            column: "id",
            type: Number,
            value: (objectData: ExportDisruptionData) => objectData.id,
        },
        {
            column: "title",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.title,
        },
        {
            column: "mode",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.mode,
        },
        {
            column: "operator wide",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData["operator wide"],
        },
        {
            column: "network wide",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData["network wide"],
        },
        {
            column: "services affected",
            type: Number,
            value: (objectData: ExportDisruptionData) => objectData["services affected"],
        },
        {
            column: "stops affected",
            type: Number,
            value: (objectData: ExportDisruptionData) => objectData["stops affected"],
        },
        {
            column: "start",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.start,
        },
        {
            column: "end",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.end,
        },
        {
            column: "severity",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.severity,
        },
        {
            column: "live",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.live,
        },
        {
            column: "status",
            type: String,
            value: (objectData: ExportDisruptionData) => objectData.severity,
        },
    ];
};
