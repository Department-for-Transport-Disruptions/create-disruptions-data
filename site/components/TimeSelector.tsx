import React, { ReactElement } from 'react';
import { ErrorInfo } from '../interfaces';
import FormElementWrapper from './FormElementWrapper';

interface TimeSelectorProps {
    inputs?: {
        hourInput: string;
        minuteInput: string;
    };
    errors?: ErrorInfo[];
    startOrEnd: 'start' | 'end';
}

const TimeSelector = ({ inputs, startOrEnd, errors = [] }: TimeSelectorProps): ReactElement => (
        <FormElementWrapper errors={errors} errorId={`${startOrEnd}-day-input`} errorClass="govuk-date-input--error">
            <div className="govuk-date-input" id={`${startOrEnd}-date`}>
                <div className="govuk-date-input__item">
                    <div className="govuk-form-group">
                        <label className="govuk-label govuk-date-input__label" htmlFor={`${startOrEnd}-hour-input`}>
                            Hour
                        </label>
                        <input
                            className={`govuk-input govuk-date-input__input govuk-input--width-2 ${
                                errors.length > 0 ? 'govuk-input--error' : ''
                            } `}
                            id={`${startOrEnd}-hour-input`}
                            name={`${startOrEnd}Hour`}
                            type="text"
                            defaultValue={inputs?.hourInput}
                        />
                    </div>
                </div>
                <div className="govuk-date-input__item">
                    <div className="govuk-form-group">
                        <label className="govuk-label govuk-date-input__label" htmlFor={`${startOrEnd}-minute-input`}>
                            Minute
                        </label>
                        <input
                            className={`govuk-input govuk-date-input__input govuk-input--width-2 ${
                                errors.length > 0 ? 'govuk-input--error' : ''
                            } `}
                            id={`${startOrEnd}-minute-input`}
                            name={`${startOrEnd}Minute`}
                            type="text"
                            defaultValue={inputs?.minuteInput}
                        />
                    </div>
                </div>
            </div>
        </FormElementWrapper>
);

export default TimeSelector;
