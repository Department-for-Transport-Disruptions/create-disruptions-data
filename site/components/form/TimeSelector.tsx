import { Dispatch, ReactElement, SetStateAction } from "react";
import FormElementWrapper, { FormGroupWrapper } from "./FormElementWrapper";
import { ErrorInfo } from "../../interfaces";
import { PageState, PageInputs } from "../../pages/create-disruption";

interface TimeSelectorProps {
    header: string;
    input?: string;
    errors?: ErrorInfo[];
    disabled: boolean;
    inputId: string;
    inputName: string;
    pageState: PageState;
    hint?: {
        id: string;
        text: string;
    };
    updatePageState: Dispatch<SetStateAction<PageState>>;
    updaterFunction: (
        currentState: PageState,
        setPageState: Dispatch<SetStateAction<PageState>>,
        inputName: keyof PageInputs,
        input: string | Date | null,
        error?: ErrorInfo,
    ) => void;
}

const TimeSelector = ({
    header,
    input,
    disabled,
    inputId,
    inputName,
    pageState,
    hint,
    updatePageState,
    updaterFunction,
}: TimeSelectorProps): ReactElement => {
    return (
        <FormGroupWrapper errorIds={[inputId]} errors={pageState.errors}>
            <fieldset className="govuk-fieldset mb-7.5" role="group" aria-describedby={hint ? hint.id : undefined}>
                <legend className="govuk-fieldset__legend">
                    <h3 className="govuk-heading-s govuk-!-margin-bottom-0">{header}</h3>
                </legend>
                {hint ? (
                    <div id={hint.id} className="govuk-hint">
                        {hint.text}
                    </div>
                ) : null}

                <FormElementWrapper errors={pageState.errors} errorId={inputId} errorClass="govuk-input--error">
                    <input
                        className="govuk-input govuk-date-input__input govuk-input--width-4"
                        id={inputId}
                        name={inputName}
                        type="text"
                        defaultValue={input}
                        disabled={disabled}
                        placeholder={disabled ? "N/A" : "hhmm"}
                        onBlur={(e) => {
                            const input = e.target.value;
                            if (!input) {
                                updaterFunction(pageState, updatePageState, inputId as keyof PageInputs, input, {
                                    id: inputId,
                                    errorMessage: "Enter a time in hhmm format",
                                });
                            } else {
                                updaterFunction(pageState, updatePageState, inputId as keyof PageInputs, input);
                            }
                        }}
                    />
                </FormElementWrapper>
            </fieldset>
        </FormGroupWrapper>
    );
};

export default TimeSelector;
