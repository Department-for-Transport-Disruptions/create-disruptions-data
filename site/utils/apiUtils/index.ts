import { Consequence } from "@create-disruptions-data/shared-ts/disruptionTypes";
import { MAX_CONSEQUENCES } from "@create-disruptions-data/shared-ts/disruptionTypes.zod";
import { SocialMediaPostStatus } from "@create-disruptions-data/shared-ts/enums";
import { NextApiRequest, NextApiResponse } from "next";
import { parseCookies, setCookie } from "nookies";
import { TwitterApi } from "twitter-api-v2";
import { z } from "zod";
import { IncomingMessage, ServerResponse } from "http";
import { notEmpty } from "..";
import {
    COOKIES_POLICY_COOKIE,
    COOKIE_CSRF,
    COOKIES_ID_TOKEN,
    COOKIE_PREFERENCES_COOKIE,
    DISRUPTION_DETAIL_PAGE_PATH,
    REVIEW_DISRUPTION_PAGE_PATH,
    COOKIES_REFRESH_TOKEN,
} from "../../constants";
import { upsertConsequence } from "../../data/dynamo";
import { getAccessToken, publishToHootsuite } from "../../data/hootsuite";
import { getAuthedTwitterClient, sendTweet } from "../../data/twitter";
import { TooManyConsequencesError } from "../../errors";
import { PageState } from "../../interfaces";
import { SocialMediaPost } from "../../schemas/social-media.schema";
import logger from "../logger";

export const setCookieOnResponseObject = (
    cookieName: string,
    cookieValue: string,
    res: NextApiResponse | ServerResponse<IncomingMessage>,
    lifetime?: number,
    httpOnly = true,
    sameSite: "lax" | "strict" | "none" = "lax",
): void => {
    // From docs: All cookies are httponly by default, and cookies sent over SSL are secure by
    // default. An error will be thrown if you try to send secure cookies over an insecure socket.
    setCookie({ res }, cookieName, cookieValue, {
        path: "/",
        sameSite,
        secure: process.env.NODE_ENV !== "development",
        maxAge: lifetime,
        httpOnly,
    });
};

export const destroyCookieOnResponseObject = (
    cookieName: string,
    res: NextApiResponse | ServerResponse<IncomingMessage>,
): void => {
    setCookieOnResponseObject(cookieName, "", res, 0);
};

export const redirectTo = (res: NextApiResponse | ServerResponse, location: string): void => {
    res.writeHead(302, {
        Location: location,
    });
    res.end();
};

export const redirectToError = (
    res: NextApiResponse | ServerResponse,
    message?: string,
    context?: string,
    error?: Error,
): void => {
    if (message && context && error) {
        logger.error(message, { context, error: error.stack });
    }

    redirectTo(res, "/500");
};

export const cleardownCookies = (req: NextApiRequest, res: NextApiResponse) => {
    const cookies = parseCookies({ req });

    const saveList = [
        COOKIES_POLICY_COOKIE,
        COOKIE_PREFERENCES_COOKIE,
        COOKIES_ID_TOKEN,
        COOKIES_REFRESH_TOKEN,
        COOKIE_CSRF,
    ];

    Object.keys(cookies).forEach((cookie) => {
        if (!saveList.includes(cookie)) {
            destroyCookieOnResponseObject(cookie, res);
        }
    });
};

export const getPageState = <T>(errorCookie: string, schemaObject: z.ZodType<T>, disruptionId?: string, data?: T) => {
    const inputsProps: PageState<Partial<T>> = {
        errors: [],
        inputs: {},
        disruptionId: disruptionId || "",
    };

    if (errorCookie) {
        return { ...(JSON.parse(errorCookie) as PageState<Partial<T>>), disruptionId: disruptionId || "" };
    }

    if (disruptionId) {
        const parsedData = schemaObject.safeParse(data);

        if (parsedData.success) {
            inputsProps.inputs = parsedData.data;
        }
    }

    return inputsProps;
};

export const getReturnPage = (req: NextApiRequest) => {
    const queryParam = req.headers.referer?.split("?")[1];
    const decodedQueryParam = queryParam ? decodeURIComponent(queryParam) : null;
    return decodedQueryParam?.includes(REVIEW_DISRUPTION_PAGE_PATH) ||
        decodedQueryParam?.includes(DISRUPTION_DETAIL_PAGE_PATH)
        ? queryParam
        : null;
};

export const isDisruptionFromTemplate = (req: NextApiRequest) => {
    const queryParam = req.headers.referer?.split("?")[1];

    const decodedQueryParam = queryParam ? decodeURIComponent(queryParam) : null;
    return decodedQueryParam?.includes(DISRUPTION_DETAIL_PAGE_PATH) && decodedQueryParam.includes("template=true")
        ? queryParam
        : null;
};

export const delay = (ms: number) => new Promise((res) => setTimeout(res, ms));

export const publishSocialMedia = async (
    socialMediaPosts: SocialMediaPost[],
    orgId: string,
    isUserStaff: boolean,
    canPublish: boolean,
) => {
    const authedTwitterClients: Record<string, TwitterApi> = {};
    const hootsuiteAccessTokens: Record<string, string> = {};

    const uniqueTwitterSocialAccounts = new Set(
        socialMediaPosts.filter((post) => post.accountType === "Twitter").map((post) => post.socialAccount),
    );

    const uniqueHootsuiteSocialAccounts = new Set(
        socialMediaPosts.filter((post) => post.accountType === "Hootsuite").map((post) => post.socialAccount),
    );

    for (const socialAccount of uniqueHootsuiteSocialAccounts) {
        const accessToken = await getAccessToken(orgId, socialAccount);
        hootsuiteAccessTokens[socialAccount] = accessToken;
    }

    for (const socialAccount of uniqueTwitterSocialAccounts) {
        const authedClient = await getAuthedTwitterClient(orgId, socialAccount);

        if (authedClient) {
            authedTwitterClients[socialAccount] = authedClient;
        }
    }

    const socialMediaPromises = socialMediaPosts.map((post) => {
        if (post.status === SocialMediaPostStatus.pending) {
            if (post.accountType === "Twitter") {
                return sendTweet(orgId, post, isUserStaff, canPublish, authedTwitterClients[post.socialAccount]);
            }

            return publishToHootsuite(post, orgId, isUserStaff, canPublish, hootsuiteAccessTokens[post.socialAccount]);
        }

        return null;
    });

    await Promise.all(socialMediaPromises);
};

export const redirectToWithQueryParams = (
    req: NextApiRequest,
    res: NextApiResponse,
    queryParamsToForward: string[],
    location: string,
    paramsToAdd?: string[],
) => {
    const queryStringParams = queryParamsToForward
        .map((p) => {
            const paramValue = req.query[p];

            return paramValue ? `${p}=${paramValue.toString()}` : null;
        })
        .filter(notEmpty);

    redirectTo(
        res,
        `${location}${
            [...queryStringParams, ...(paramsToAdd || [])].length > 0
                ? `?${[...queryStringParams, ...(paramsToAdd || [])].join("&")}`
                : ""
        }`,
    );
};

export const handleUpsertConsequence = async (
    consequence: Consequence | Pick<Consequence, "disruptionId" | "consequenceIndex">,
    orgId: string,
    isOrgStaff: boolean,
    isTemplate: boolean,
    inputs: unknown,
    errorCookie: string,
    res: NextApiResponse,
) => {
    try {
        return await upsertConsequence(consequence, orgId, isOrgStaff, isTemplate);
    } catch (e) {
        if (e instanceof TooManyConsequencesError) {
            setCookieOnResponseObject(
                errorCookie,
                JSON.stringify({
                    inputs,
                    errors: [
                        {
                            id: "",
                            errorMessage: `Max consequence limit of ${MAX_CONSEQUENCES} has been reached`,
                        },
                    ],
                }),
                res,
            );
        }

        throw e;
    }
};
