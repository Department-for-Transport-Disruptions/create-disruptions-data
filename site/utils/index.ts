import { ErrorInfo } from "interfaces";
import { ServerResponse } from "http";
import { NextApiResponse } from "next";

export const buildTitle = (errors: ErrorInfo[], title: string): string => {
    if (errors.length > 0) {
        return `Error: ${title}`;
    }

    return title;
};

export const redirectTo = (res: NextApiResponse | ServerResponse, location: string): void => {
    res.writeHead(302, {
        Location: location,
    });
    res.end();
};
