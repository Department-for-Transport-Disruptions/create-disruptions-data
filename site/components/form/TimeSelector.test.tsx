import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { describe, it, expect, vi } from "vitest";
import { z } from "zod";
import TimeSelector from "./TimeSelector";
import { TestInputs } from "../../interfaces";
import { setZodDefaultError } from "../../utils";

describe("TimeSelector", () => {
    it("should render correctly with no inputs", () => {
        const tree = renderer
            .create(
                <TimeSelector<TestInputs>
                    display="Start time"
                    value={""}
                    disabled={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with inputs", () => {
        const tree = renderer
            .create(
                <TimeSelector<TestInputs>
                    display="Start time"
                    value={"0900"}
                    disabled={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with errors", () => {
        const tree = renderer
            .create(
                <TimeSelector<TestInputs>
                    display="Start time"
                    hint="Test Hint"
                    value={"three thirty"}
                    initialErrors={[{ errorMessage: "There was an error", id: "disruption-reason" }]}
                    disabled={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly when disabled", () => {
        const tree = renderer
            .create(
                <TimeSelector<TestInputs>
                    display="Start time"
                    value={"three thirty"}
                    disabled
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should validate minLength and display error", async () => {
        const { unmount } = render(
            <TimeSelector<TestInputs>
                display="Start time"
                hint="Test Hint"
                value={""}
                disabled={false}
                inputName="field1"
                stateUpdater={vi.fn()}
                initialErrors={[]}
                schema={z.string(setZodDefaultError("Error: Test Error")).min(1)}
            />,
        );

        await userEvent.click(screen.getByLabelText("Start time"));
        await userEvent.tab();

        expect(screen.getByText("Error: Test Error")).toBeTruthy();

        unmount();
    });

    it("should not display error if the component is disabled", async () => {
        const { unmount } = render(
            <TimeSelector<TestInputs>
                display="Start time"
                hint="Test Hint"
                value={""}
                disabled
                inputName="field1"
                stateUpdater={vi.fn()}
                schema={z.string(setZodDefaultError("Error: Test Error")).min(1)}
            />,
        );

        await userEvent.click(screen.getByLabelText("Start time"));
        await userEvent.tab();

        expect(screen.queryByText("Error: Test Error")).toBeFalsy();

        unmount();
    });
});
