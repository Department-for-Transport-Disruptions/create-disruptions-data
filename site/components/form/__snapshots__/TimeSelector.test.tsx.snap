// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimeSelector > should render correctly when disabled 1`] = `
<div
  className="govuk-form-group"
>
  <fieldset
    aria-describedby="publish-start-time-hint"
    className="govuk-fieldset mb-7.5"
    role="group"
  >
    <legend
      className="govuk-fieldset__legend"
    >
      <h3
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        What is the start time?
      </h3>
    </legend>
    <div
      className="govuk-hint"
      id="publish-start-time-hint"
    >
      Enter the time in 24hr format. For example 0900 is 9am, 1730 is 5:30pm
    </div>
    <div
      className=""
    >
      <input
        className="govuk-input govuk-date-input__input govuk-input--width-4"
        disabled={true}
        id="publish-start-time-input"
        name="publishStartTime"
        onBlur={[Function]}
        placeholder="N/A"
        type="text"
      />
    </div>
  </fieldset>
</div>
`;

exports[`TimeSelector > should render correctly with errors 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <fieldset
    aria-describedby="publish-start-time-hint"
    className="govuk-fieldset mb-7.5"
    role="group"
  >
    <legend
      className="govuk-fieldset__legend"
    >
      <h3
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        What is the start time?
      </h3>
    </legend>
    <div
      className="govuk-hint"
      id="publish-start-time-hint"
    >
      Enter the time in 24hr format. For example 0900 is 9am, 1730 is 5:30pm
    </div>
    <div
      className=""
    >
      <span
        className="govuk-error-message"
        id="publish-start-time-input-error"
      >
        <span
          className="govuk-visually-hidden"
        >
          Error: 
        </span>
        Enter a time in hhmm format
      </span>
      <input
        aria-describedby="publish-start-time-input-error"
        className="govuk-input govuk-date-input__input govuk-input--width-4 govuk-input--error"
        defaultValue=""
        disabled={false}
        id="publish-start-time-input"
        name="publishStartTime"
        onBlur={[Function]}
        placeholder="hhmm"
        type="text"
      />
    </div>
  </fieldset>
</div>
`;

exports[`TimeSelector > should render correctly with inputs 1`] = `
<div
  className="govuk-form-group"
>
  <fieldset
    aria-describedby="publish-start-time-hint"
    className="govuk-fieldset mb-7.5"
    role="group"
  >
    <legend
      className="govuk-fieldset__legend"
    >
      <h3
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        What is the start time?
      </h3>
    </legend>
    <div
      className="govuk-hint"
      id="publish-start-time-hint"
    >
      Enter the time in 24hr format. For example 0900 is 9am, 1730 is 5:30pm
    </div>
    <div
      className=""
    >
      <input
        className="govuk-input govuk-date-input__input govuk-input--width-4"
        defaultValue="0900"
        disabled={false}
        id="publish-start-time-input"
        name="publishStartTime"
        onBlur={[Function]}
        placeholder="hhmm"
        type="text"
      />
    </div>
  </fieldset>
</div>
`;

exports[`TimeSelector > should render correctly with no inputs 1`] = `
<div
  className="govuk-form-group"
>
  <fieldset
    aria-describedby="publish-start-time-hint"
    className="govuk-fieldset mb-7.5"
    role="group"
  >
    <legend
      className="govuk-fieldset__legend"
    >
      <h3
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        What is the start time?
      </h3>
    </legend>
    <div
      className="govuk-hint"
      id="publish-start-time-hint"
    >
      Enter the time in 24hr format. For example 0900 is 9am, 1730 is 5:30pm
    </div>
    <div
      className=""
    >
      <input
        className="govuk-input govuk-date-input__input govuk-input--width-4"
        disabled={false}
        id="publish-start-time-input"
        name="publishStartTime"
        onBlur={[Function]}
        placeholder="hhmm"
        type="text"
      />
    </div>
  </fieldset>
</div>
`;
